### PR TITLE
Mark GenericByteArray::new_unchecked unsafe

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Review by jackwener — APPROVED** [review]

File: arrow-array/src/array/byte_array.rs [hunk 1]
```diff
@@ -159,7 +159,7 @@ impl<T: ByteArrayType> GenericByteArray<T> {
     /// # Safety
     ///
     /// Safe if [`Self::try_new`] would not error
-    pub fn new_unchecked(
+    pub unsafe fn new_unchecked(
```
Comment (findepi) [inline arrow-array/src/array/byte_array.rs:162]: Can you please elaborate more on the why we need `unsafe` here?

cc @Xuanwo

Comment (tustvold) [reply arrow-array/src/array/byte_array.rs:162]: See the checks done by try_new

Comment (findepi) [reply arrow-array/src/array/byte_array.rs:162]: thanks @tustvold . so this allows creating an array without validation, thus with potentially invalid data.  is this also unsafe in rust unsafe sense (potential program undefined behavior)?

Comment (tustvold) [reply arrow-array/src/array/byte_array.rs:162]: Yes, very much so. Many kernels and methods make assumptions about data validity.

**Pull request merged**